eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs

```diff
@@ -262,59 +266,79 @@ roundupBytesToWords b = quot (b + wordLength - 1) wordLength
 -- Maybe better to make this distinction in the TxOut de/serialization
 
 decodeValue ::
+  forall c s.
   Crypto c =>
   Decoder s (MaryValue c)
-decodeValue = do
-  tt <- peekTokenType
-  case tt of
-    TypeUInt -> inject . Coin <$> decodeInteger
-    TypeUInt64 -> inject . Coin <$> decodeInteger
-    TypeNInt -> inject . Coin <$> decodeInteger
-    TypeNInt64 -> inject . Coin <$> decodeInteger
-    TypeListLen -> decodeValuePair decodeInteger
-    TypeListLen64 -> decodeValuePair decodeInteger
-    TypeListLenIndef -> decodeValuePair decodeInteger
-    _ -> fail $ "Value: expected array or int, got " ++ show tt
+decodeValue =
+  ifDecoderVersionAtLeast
+    (natVersion @9)
+    (decode' conwayCoin)
+    (decode' preConwayCoin)
+  where
+    conwayCoin :: Decoder s (MaryValue c)
+    conwayCoin = inject . Coin . fromInteger <$> decodePositiveInteger64
+    preConwayCoin :: Decoder s (MaryValue c)
+    preConwayCoin = inject . word64ToCoin <$> decodeWord64
+    decode' :: Decoder s (MaryValue c) -> Decoder s (MaryValue c)
+    decode' decodeCoin = do
+      tt <- peekTokenType
+      case tt of
+        TypeUInt -> decodeCoin
+        TypeUInt64 -> decodeCoin
+        TypeListLen -> decodeValuePair decodeInteger
+        TypeListLen64 -> decodeValuePair decodeInteger
+        TypeListLenIndef -> decodeValuePair decodeInteger
+        _ -> fail $ "Value: expected array or int, got " ++ show tt
 
 decodeValuePair ::
   Crypto c =>
   (forall t. Decoder t Integer) ->
   Decoder s (MaryValue c)
-decodeValuePair decodeAmount =
+decodeValuePair decoder =
   decode $
     RecD MaryValue
-      <! D decodeAmount
-      <! D (decodeMultiAssetMaps decodeAmount)
+      <! D decoder
+      <! D decodeMultiAssetValue
 
 encodeMultiAssetMaps ::
   Crypto c =>
   MultiAsset c ->
   Encoding
 encodeMultiAssetMaps (MultiAsset m) = toCBOR m
 
-decodeMultiAssetMaps ::
+-- | MultiAsset can be of two types: Value and Mint.
```

Comment:
There aren't any Haskell type `Mint`, so this comment can be a bit confusing, maybe instead
```suggestion
-- | `MultiAsset` can be used in two different circumstances: during minting and in `MaryValue` while sending.
-- In latter case MultiAsset amounts cannot be negative, while in minting negative indicates burning.
```